litex/build/quicklogic/f4pga.py

```diff
@@ -47,48 +48,59 @@ def build_io_constraints(self):
-        # build bit file (default)
-        makefile.append(f"build/{self._build_name}.bit:")
-        makefile.append("\tql_symbiflow -compile -d {device} -P {part} -v {verilog} -t {top} -p {pcf}".format(
-            device  = self.platform.device,
-            part    = {"ql-eos-s3": "PU64"}.get(self.platform.device),
-            verilog = f"{self._build_name}.v",
-            top     = self._build_name,
-            pcf     = f"{self._build_name}.pcf"
-        ))
-        # build header to include in CPU firmware
-        makefile.append("{top}_bit.h: build/{top}.bit".format(top=self._build_name))
-        makefile.append(f"\t(cd build; TOP_F=$(TOP_F) symbiflow_write_bitheader)")
-        # build binary to write in dedicated FLASH area
-        makefile.append("{top}.bin: build/{top}.bit".format(top=self._build_name))
-        makefile.append(f"\t(cd build; TOP_F=$(TOP_F) symbiflow_write_binary)")
-
-        # Generate Makefile.
-        tools.write_to_file("Makefile", "\n".join(makefile))
-
-        return "Makefile"
+        part = {"ql-eos-s3": "PU64"}.get(self.platform.device)
+        flow = {
+            "default_part": "EOS3FF512-PDN64",
+            "values": {
+                "top": self._build_name
+            },
+            "dependencies": {
+                "sources": [
+                    f"{self._build_name}.v"
+                ],
+                "synth_log": "synth.log",
+                "pack_log": "pack.log",
+                "analysis_log": "analysis.log"
+            },
+            "EOS3FF512-PDN64": {
+                "default_target": "bitstream",
+                "dependencies": {
+                    "build_dir": "build",
+                    "pcf": f"{self._build_name}.pcf",
+                    "sdc-in": f"{self._build_name}.sdc"
+                },
+                "values": {
+                    "part": self.platform.device,
+                    "package": part
+                }
+            }
+        };
+
+        tools.write_to_file("flow.json", json.dumps(flow))
+        return "flow.json"
 
     def run_script(self, script):
-        make_cmd = ["make", "-j1"]
-
-        if which("ql_symbiflow") is None:
-            msg = "Unable to find QuickLogic Symbiflow toolchain, please:\n"
-            msg += "- Add QuickLogic Symbiflow toolchain to your $PATH."
-            raise OSError(msg)
+        make_cmd = ["f4pga", "-vvv", "build", "--flow", "flow.json"]
 
         if subprocess.call(make_cmd) != 0:
             raise OSError("Error occured during QuickLogic Symbiflow's script execution.")
+
+        make_cmd.append("--target")
+        if subprocess.call(make_cmd + ["bitstream_openocd"]) != 0:
+            raise OSError("Error occured during QuickLogic Symbiflow's script execution.")
+
+        if subprocess.call(make_cmd + ["bitstream_jlink"]) != 0:
+            raise OSError("Error occured during QuickLogic Symbiflow's script execution.")
```

Comment:
These two commands produces only bitstream/rules for *openOCD* and *jlink* so we lost ability to obtain *.bit.bin* and *.bit.h* files: it's required to add `bitstream_bitheader` and `bitstream_binary` too.
What do you think about using a dict to iterate on all 4 targets and improving error message with target name?